File: src/lexer/token.rs
use owo_colors::OwoColorize;

use crate::lexer::Location;

#[derive(Debug, Clone)]
pub struct Token {
    pub kind: TokenKind,
    pub loc: Location,
}

impl std::fmt::Display for Token {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "{}: {}", self.loc, self.kind)
    }
}

impl Token {
    pub fn new(kind: TokenKind, loc: Location) -> Self {
        Self { kind, loc }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum TokenKind {
    // keywords
    Func,
    Defer,
    Return,

    // literals
    Identifier(String),
    Number(String),

    // arithmetic
    Plus,
    Minus,
    Slash,
    Star,

    // logical
    Not,
    And,
    Or,

    // comparative
    Less,
    LessEqual,
    Greater,
    GreaterEqual,
    DoubleEqual,

    // assignment
    Colon,
    DoubleColon,
    Equal,

    // other
    Comma,
    LeftParen,
    RightParen,
    LeftCurly,
    RightCurly,
    Eof,
}

impl std::fmt::Display for TokenKind {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        match self {
            TokenKind::Identifier(name) => {
                write!(f, "{}({:?})", "Identifier".cyan(), name)
            }
            TokenKind::Number(value) => {
                write!(f, "{}({:?})", "Number".cyan(), value)
            }
            _ => write!(f, "{:?}", self.cyan()),
        }
    }
}

#[derive(Clone)]
pub struct TokenList {
    tokens: Vec<Token>,
}

impl std::fmt::Display for TokenList {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        for token in &self.tokens {
            write!(f, "{}\n", token)?;
        }
        Ok(())
    }
}

impl TokenList {
    pub fn new() -> Self {
        Self { tokens: Vec::new() }
    }

    pub fn push(&mut self, token: Token) {
        self.tokens.push(token);
    }

    pub fn len(&self) -> usize {
        self.tokens.len()
    }

    pub fn to_vec(&self) -> Vec<Token> {
        self.tokens.clone()
    }
}

File: src/lexer/Token.zig
const std = @import("std");
const StringPool = @import("../root.zig").StringPool;

pub const keywords = std.StaticStringMap(Tag).initComptime(.{
    .{ "mut", .mut },
    .{ "if", .@"if" },
    .{ "else", .@"else" },
    .{ "func", .func },
    .{ "import", .import },
    .{ "return", .@"return" },
});

tag: Tag,
start: Index,
str_id: StringPool.ID,

pub const Index = u32;

pub const Tag = enum {
    // complex tokens
    identifier,
    int,
    float,
    string,

    // keywords
    mut,
    @"if",
    @"else",
    @"and",
    @"or",
    func,
    import,
    @"return",

    // single-char tokens
    plus,
    minus,
    star,
    slash,
    bang,
    colon,
    comma,
    dot,
    equal,
    left_paren,
    right_paren,
    left_bracket,
    right_bracket,
    left_curly,
    right_curly,

    // double-char tokens
    double_colon,

    // call-conv tokens
    cc_honey,
    cc_c,

    // special tokens
    newline,
    invalid,
    eof,
};
